lib/src/rules/avoid_runtimeType_toString.dart

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Avoid calling toString() on runtimeType.';
+
+const _details = r'''
+
+Calling `toString` on a runtime type is a non-trivial operation that can
+negatively impact performance. It's better to avoid it.
+
+**BAD:**
+```
+class A {
+  String toString() => '$runtimeType()';
+}
+```
+
+**GOOD:**
+```
+class A {
+  String toString() => 'A()';
+}
+```
+
+This lint has some exceptions where performance is not a problem or where real
+type information is more important than performance:
+
+* in assertion
+* in throw expressions
+* in catch clauses
+* in mixin declaration
+* in abstract class
+
+''';
+
+class AvoidRuntimeTypeToString extends LintRule implements NodeLintRule {
+  AvoidRuntimeTypeToString()
+      : super(
+            name: 'avoid_runtimeType_toString',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this);
+    registry.addInterpolationExpression(this, visitor);
+    registry.addMethodInvocation(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    if (_canSkip(node)) {
+      return;
+    }
+    if (node.methodName.name == 'toString' &&
+        _isRuntimeTypeAccess(node.target)) {
```

Comment:
It probably isn't common enough to worry about, but this should really be `node.realTarget` in order to handle cascades.

Reply:
No, me either. Might not be worth it in this case. That said, I've seen users do lots of things I didn't imagine they'd do. :-)